### ISSUE TITLE
[HIGH][$250] Send invoice - Category is missing in confirmation page but it is present in invoice view

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#41281]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.68-0
**Reproducible in staging?:** Yes
**Reproducible in production?:** No (New feature)
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause Internal Team
**Slack conversation:**

## Action Performed:
Pre condition:Create a new workspace and use to send Invoice
1. Go to staging.new.expensify.com
2. Go to FAB > Send invoice.
3. Enter amount and select a participant.
4. Note that "Category" is not present in the confirmation page.
5. Create the invoice.
6. Click on the preview in workspace chat.
## Expected Result:
"Category" row should be present in the confirmation page if it is present in the invoice details page.

## Actual Result:
"Category" row is not present in the confirmation page when it is present in the invoice details page.

## Workaround:
Unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [X] Android: Native
- [X] Android: mWeb Chrome
- [X] iOS: Native
- [X] iOS: mWeb Safari
- [X] MacOS: Chrome / Safari
- [X] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/6edfd22c-ecbe-4cf2-b55d-d8e9f97cf5fe

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~017a0bd1ec84ee8690</li>
        <li>Upwork Job ID: 1785394941425930240</li>
        <li>Last Price Increase: 2024-08-07</li>
<li>Automatic offers: </li>
<ul>
<li>Pujan92 | Reviewer | 0</li>
<li>gijoe0295 | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @youssef-lr </details>

**Comment by m-natarajan** [comment]
We think this bug might be related to #wave-collect - Release 1

**Comment by m-natarajan** [comment]
@sonialiap FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors.

**Comment by robertjchen** [comment]
Does not appear to be a serious deploy blocker to me and can be addressable by external contributors 👍

**Comment by danielrvidal** [comment]
Given this is core functionality for anyone with a workspace, I'm marking it as CRITICAL as it might be needed in the demos.

**Comment by danielrvidal** [comment]
@Pujan92 This is a high-priority one. Once you get proposals, can you make sure to accelerate review to get this picked up? Thank you!

**Comment by brunovjk** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Inconsistency in the 'Category' field when creating and editing 'Send Invoice'

### What is the root cause of that problem?
When we send 'Send Invoice' to a user, the 'category' field appears in the 'MoneyRequestView'

### What changes do you think we should make in order to solve the problem?
With this we must 'remove' the 'category' field from 'MoneyRequestView' if it is not `isPolicyExpenseChat`. Something like:
```
    const isPolicyExpenseChat = ReportUtils.isPolicyExpenseChat(report);
```
Instead of the current one:
https://github.com/Expensify/App/blob/54d971a5a2a1155544a5e5d15b2e0103a82e8b75/src/components/ReportActionItem/MoneyRequestView.tsx#L168
### What alternative solutions did you explore? (Optional)
N/A

**Comment by brunovjk** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Inconsistency in the display of the "Category" field when creating and viewing `Send invoice` request.

### What is the root cause of that problem?
Currently, the "Category" field may appear when viewing an invoice (in the `MoneyRequestView`) even though it does not reflect an actual policy expense.

Root cause lies in the conditional logic determining when to display the "Category" field. The current implementation seems to be based on `ReportUtils.isReportInGroupPolicy(report)`:
https://github.com/Expensify/App/blob/54d971a5a2a1155544a5e5d15b2e0103a82e8b75/src/components/ReportActionItem/MoneyRequestView.tsx#L168
This condition may need refinement to accurately filter out non-policy expenses.

### What changes do you think we should make in order to solve the problem?

I propose we modify the logic controlling the "Category" field's visibility within the `MoneyRequestView`. Specifically, we can introduce the following change:
```
const isPolicyExpenseChat = ReportUtils.isPolicyExpenseChat(report);
```

This refined condition will ensure that the "Category" field is displayed *only* when the invoice truly represents a policy expense.

**Comment by Pujan92** [comment]
@cristipaval @danielrvidal Needs confirmation regarding the expected result, Categories should be displayed for the non-policy expense chat invoices or not? Bcoz we usually show it for the `policyExpenseChat` only.

I am more towards @brunovjk's [proposal](https://github.com/Expensify/App/issues/41281#issuecomment-2087363366) where they suggested using the consistent condition in both places and in this case, it won't show categories in `MoneyRequestView` as the rootParent Report isn't of type `policyExpenseChat`.

**Comment by Pujan92** [comment]
> In invoice confirmation step, we used a "optimistic" report, thus it is {} causing isPolicyExpenseChat to be false:

@gijoe0295 wondering why the issue is not there for the submit expense.

**Comment by brunovjk** [comment]
> confirmation regarding the expected result, Categories should be displayed for the non-policy expense chat invoices or not?

I wonder the same.

I think the problem here is about inconsistency:
https://github.com/Expensify/App/blob/573d9893ae2a8e6e4ebe64088cb2fbf62ce81804/src/components/ReportActionItem/MoneyRequestView.tsx#L168
https://github.com/Expensify/App/blob/573d9893ae2a8e6e4ebe64088cb2fbf62ce81804/src/pages/iou/request/step/IOURequestStepConfirmation.tsx#L137
We need to align this. Change variable name if necessary, add/remove `useMemo` if necessary, ... .

**Comment by Pujan92** [comment]
Makes sense @gijoe0295. Your solution seems to work where we need to use `policy.id` in case the report is not available. Still, I am tagging @VickyStash to get their input on any other straightforward alternative.

**Comment by brunovjk** [comment]
> confirmation regarding the expected result, Categories should be displayed for the non-policy expense chat invoices or not?

Exactly.

-------

I think the problem here is about inconsistency
https://github.com/Expensify/App/blob/99017956d92998f3e301d6a4124669a9e5e44d93/src/components/ReportActionItem/MoneyRequestView.tsx#L168
https://github.com/Expensify/App/blob/99017956d92998f3e301d6a4124669a9e5e44d93/src/pages/iou/request/step/IOURequestStepConfirmation.tsx#L137

**Comment by danielrvidal** [comment]
@Pujan92 the expected behavior is that the individual sending the invoice would see the category field as the invoice is owned by their workspace. But the individual receiving the invoice should not. Does that make sense? Let me know if I can answer anything else.

**Comment by Pujan92** [comment]
Thanks @danielrvidal.

We can proceed with @gijoe0295's [proposal](https://github.com/Expensify/App/issues/41281#issuecomment-2087750193) as their solution will fix the issue. cc: @cristipaval 

🎀👀🎀 C+ reviewed

**Comment by brunovjk** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/41281#issuecomment-2087363366)

Standardize how we handle `isPolicyExpenseChat` across both creating and viewing reports, ensuring consistent behavior throughout.

**Comment by brunovjk** [comment]
@Pujan92 I was updating my proposal, can you give me some feedback there? Thank you very much.

**Comment by danielrvidal** [comment]
@gijoe0295 can you make sure to prioritize this one tomorrow? We need this to be fixed asap so if it could be deployed to staging tomorrow that would be best. Thank you!

**Comment by Pujan92** [comment]
@gijoe0295 will you raise a PR soon or facing any issues?

**Comment by Pujan92** [comment]
Thanks :)

**Comment by Pujan92** [comment]
> However, seems like the BE currently does not support these params in `SendInvoice` API. I tried to enable and set those fields in the FE, the payload was correct but `SendInvoice` just returned empty for those fields.

@cristipaval Can you plz confirm whether the `SendInvoice` API is accepting these extra params(category, tax, ..) or not?

**Comment by Pujan92** [comment]
Ah, Ok. As all those fields appear when `isPolicyExpenseChat` is true we need to apply this logic to all components. 
Better if you share the payload here which @cristipaval can check and confirm with the backend.

**Comment by davidcardoza** [comment]
The `tag`, `taxRate`, and `taxCode` parameters can be left blank. These are parameters that we display in the UI of NewDot.

**Comment by davidcardoza** [comment]
Got it. Yeah, we will want to provide the ability for users who create an invoice to add and edit tags and taxes rates when they become available in the product. So I agree with your approach.

**Comment by davidcardoza** [comment]
Sorry to clarify, we shouldn't wait for the changes in BE. We should temporarily disable the fields mentioned. We can re-enable them when they become available.

**Comment by Pujan92** [comment]
> Sorry to clarify, we shouldn't wait for the changes in BE. We should temporarily disable the fields mentioned. We can re-enable them when they become available.

@cristipaval @davidcardoza As @gijoe0295 mentioned [here](https://github.com/Expensify/App/pull/41524#issuecomment-2092759838), in case we disable those fields(tag, tax) then it may show the RBR(violation) to the user if any of them is mandatory and user can't fix that due to we are hiding/disabling those fields. 

If we can't put this on hold till the backend changes then the option which looks more feasible to me is 

- Hide these fields in the Confirmation screen of the Send invoice flow
- In the money request view show these fields and let user update as that API works

Let us know your thoughts to proceed

**Comment by davidcardoza** [comment]
>If we can't put this on hold till the backend changes then the option which looks more feasible to me is 
>
>- Hide these fields in the Confirmation screen of the Send invoice flow
>- In the money request view show these fields and let user update as that API works

This approach makes sense. Once the fields become available in the product, we will include them in this send invoice workflow and stop hiding them. However, our immediate priority is to perfect the invoice flow to meet our internal deadline. For now, let's hide these fields in the confirmation screen for the send invoice flow. In the money request view, however, we will display the fields.

**Comment by Pujan92** [comment]
Thanks @davidcardoza for the confirmation. 
@cristipaval [PR](https://github.com/Expensify/App/pull/41524) is ready for your review.

**Comment by cristipaval** [comment]
@gijoe0295 could you please quickly reopen a PR?

**Comment by sonialiap** [comment]
Payment summary
@Pujan92 Reviewer $250 - paid ✔️ - please complete the [checklist](https://github.com/Expensify/App/issues/41281#issuecomment-2106918936) :) 
@gijoe0295 Contributor $250 - paid ✔️

**Comment by cristipaval** [comment]
Making this one Weekly, because @Pujan92 is ooo

**Comment by sonialiap** [comment]
@Pujan92 looks like [you're back](https://expensify.slack.com/archives/C01GTK53T8Q/p1716901600123099?thread_ts=1716900708.700579&cid=C01GTK53T8Q), hope you had a good holiday! Could you please complete the [checklist](https://github.com/Expensify/App/issues/41281#issuecomment-2106918936) when you get the chance?

**Comment by Pujan92** [comment]
**BugZero Checklist**: The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [X] [@Pujan92] The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/40015
- [X] [@Pujan92] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: - https://github.com/Expensify/App/pull/40015#issuecomment-2149902531
 - [X] [@Pujan92] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: NA
 - [X] [@Pujan92] Determine if we should create a regression test for this bug. Yes
 - [X] [@Pujan92] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

### Regression Test Steps

1. Press FAB and select the Send invoice option
2. Select any participant and a Collect workspace with categories enabled
3. Verify in the confirmation screen that the Category menu shows and you can select a category

**Comment by sonialiap** [comment]
Thanks!

**Comment by isagoico** [comment]
Reopening this issue as this is reproducible again: 

![image](https://github.com/user-attachments/assets/02b07144-6018-4b82-a573-3fcfde0a3721)

CC @sonialiap @davidcardoza

**Comment by cristipaval** [comment]
@davidcardoza could you please update the criticality of this one?

**Comment by cristipaval** [comment]
Not overdue, Melv. We just reopened this one.

**Comment by cristipaval** [comment]
friendly bump, @davidcardoza

**Comment by davidcardoza** [comment]
Moved this issue criticality to HIGH

**Comment by cristipaval** [comment]
Added `Help Wanted` label on it to get proposals again on it.

**Comment by cristipaval** [comment]
@Pujan92 could you please review the proposed solution?

**Comment by Pujan92** [comment]
> ```const isPolicyExpenseChat = iouType === CONST.IOU.TYPE.INVOICE || participants?.some((participant) => participant.isPolicyExpenseChat;```

@gijoe0295 Instead of considering the type `invoice` report for `isPolicyExpenseChat`, how about if we bypass the `isPolicyExpenseChat` check for the `shouldShowCategories`?

`(isTypeInvoice || isPolicyExpenseChat)`

https://github.com/Expensify/App/blob/9f1faaf8813ad874c5bf93dea4e985fd84160390/src/components/MoneyRequestConfirmationList.tsx#L251

**Comment by Pujan92** [comment]
Cool, I suggested that because currently `isPolicyExpenseChat` logic varies in different places. 

@cristipaval we can proceed with @gijoe0295.

🎀👀🎀 C+ reviewed

**Comment by cristipaval** [comment]
Thanks @sonialiap! @gijoe0295, you're good to go 🚀

**Comment by sonialiap** [comment]
@sakluger I'm OOO Aug 19-30, adding leave buddy. 
Status: waiting for build

**Comment by sakluger** [comment]
The automation didn't kick in so I will pretend to be melvin 🤖 

The solution for this issue has been :rocket: deployed to production :rocket: in version 9.0.22-9 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/47474


If no regressions arise, payment will be issued on 2024-08-27. :confetti_ball:

For reference, here are some details about the assignees on this issue:
- @Pujan92 requires payment through Upwork: [Offer link](https://www.upwork.com/nx/wm/offer/102120766)
- @gijoe0295 requires payment through Upwork: [Offer link](https://www.upwork.com/nx/wm/offer/102120768)

**Comment by sakluger** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@Pujan92] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@Pujan92] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@Pujan92] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@Pujan92] Determine if we should create a regression test for this bug.
- [ ] [@Pujan92] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@sakluger] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by sakluger** [comment]
I'm going to be offline Aug 27-29 so I'm going to assign another BZ member to handle payment.

**Comment by greg-schroeder** [comment]
@Pujan92 can you take care of [the checklist](https://github.com/Expensify/App/issues/41281#issuecomment-2302523644) here? Thanks!

**Comment by Pujan92** [comment]
@greg-schroeder I completed this earlier [here](https://github.com/Expensify/App/issues/41281#issuecomment-2149904774) for the issue and it remains same. Let me know if I need to post the same here again.

**Comment by greg-schroeder** [comment]
@Pujan92 Oops, sorry, I missed that! I was just picking this up for Sasha so I am catching up here. :)

That's fine, no need to go again

**Comment by greg-schroeder** [comment]
Just a moment, here -- @Pujan92 perhaps you can help explain. It seems this was completed back in May/June, but the issue re-surfaced and the same two folks were involved as Contributor/C+.

Was this a regression, or was an entirely new solution required? Why was this project basically re-done a couple of months later? Perhaps a quick explanation would be helpful for me

**Comment by Pujan92** [comment]
@greg-schroeder Our earlier PR changes were partially modified in this [PR](https://github.com/Expensify/App/pull/41945) to fix issue https://github.com/Expensify/App/issues/41931, due to that the issue re-surfaced again. Now while checking I can see that deployblocker was somewhat connected to regression from our PR but we didn't notify and weren't aware of that. So when this issue occurred again we opened it for the external contributors to put proposals.

**Comment by greg-schroeder** [comment]
Hmm, okay. That is a bit tricky.

Let's get a few more eyes on this. Hey @cristipaval - do you mind weighing in on this one? Do you view it as a regression or a separate issue with a unique solution?

**Comment by sonialiap** [comment]
@greg-schroeder back from OOO so can take this back, feel free to stay on or unassign yourself

**Comment by sonialiap** [comment]
@cristipaval bumping Greg's question here https://github.com/Expensify/App/issues/41281#issuecomment-2311601852 regarding whether you think this is a recession or a separate issue with a unique solution?

**Comment by sonialiap** [comment]
Cristi is on parental leave. I'm going to review this tomorrow to see if we can close without his input

**Comment by sonialiap** [comment]
Apologies, I must have accidentally closed the tab and lost the issue. Thanks for the ping

**Comment by sonialiap** [comment]
@youssef-lr could you help assess whether there was a regression here? Christi is on parental leave and I'm getting lost in the code. 

This issue here was solved with PR [41723](https://github.com/Expensify/App/pull/41723/files), then your PR [41945](https://github.com/Expensify/App/pull/41945/files) changed the solution, this caused the issue to resurface and the GH issue was reopened. A new PR, [47474](https://github.com/Expensify/App/pull/47474) was pushed and needs payment. The resurfaced issue was not flagged as a regression. [Comment from the C+](https://github.com/Expensify/App/issues/41281#issuecomment-2311597700). Do you think there was a regression caused by https://github.com/Expensify/App/pull/41723/ ?

**Comment by sonialiap** [comment]
This is ready for payment, we just need to figure out whether there was a regression. @youssef-lr bumping my question from above in hopes that you can help https://github.com/Expensify/App/issues/41281#issuecomment-2352739869

**Comment by youssef-lr** [comment]
It looks like it was a regression indeed @sonialiap

**Comment by sonialiap** [comment]
Thanks for checking, Youssef! 

@Pujan92 @gijoe0295 because this latest PR was a fix for a regression on this issue, it does not qualify for additional payment. Had this been caught within the regression period then the payment would have been decreased but since it popped up a month later, you received full payment for the first fix and nothing needs to happen at this point. Thank you for quickly addressing the regression!
